lua/luasnip/util/environ.lua

```diff
@@ -1,190 +1,193 @@
-end
+local namespaces = {}
 
-function Environ:override(env, new_env)
-	for k, v in pairs(new_env) do
-		env[k] = v
+-- Namespaces allow users to define their own environmet variables
+local function _resolve_namespace_var(full_varname)
+	local parts = vim.split(full_varname, "_")
+	if #parts < 2 then
+		return nil
 	end
-end
+	local nmsp = namespaces[parts[1]]
 
--- Variables defined in https://code.visualstudio.com/docs/editor/userdefinedsnippets#_variables
+	local varname
 
--- Inherited from TextMate
-function lazy_vars.TM_FILENAME()
-	return vim.fn.expand("%:t")
+	if nmsp then
+		varname = full_varname:sub(#parts[1] + 2)
+	else
+		nmsp = namespaces[""]
+		varname = full_varname
+	end
+	return nmsp, varname
 end
 
-function lazy_vars.TM_FILENAME_BASE()
-	return vim.fn.expand("%:t:s?\\.[^\\.]\\+$??")
-end
+local Environ = {}
 
-function lazy_vars.TM_DIRECTORY()
-	return vim.fn.expand("%:p:h")
-end
-function lazy_vars.TM_FILEPATH()
-	return vim.fn.expand("%:p")
+function Environ.is_table(var_fullname)
+	local nmsp, varname = _resolve_namespace_var(var_fullname)
+	---@diagnostic disable-next-line: need-check-nil
+	return nmsp.is_table(varname)
 end
 
--- Vscode only
+function Environ:new(pos, o)
+	o = o or {}
+	setmetatable(o, self)
 
-function lazy_vars.CLIPBOARD() -- The contents of your clipboard
-	return vim.fn.getreg('"', 1, true)
-end
+	for ns_name, ns in pairs(namespaces) do
+		local eager_vars = {}
+		if ns.init then
+			eager_vars = ns.init(pos)
+		end
+		for _, eager in ipairs(ns.eager) do
+			if not eager_vars[eager] then
+				eager_vars[eager] = ns.vars(eager)
+			end
+		end
 
-local function buf_to_ws_part()
-	local LSP_WORSKPACE_PARTS = "LSP_WORSKPACE_PARTS"
-	local ok, ws_parts = pcall(vim.api.nvim_buf_get_var, 0, LSP_WORSKPACE_PARTS)
-	if not ok then
-		local file_path = vim.fn.expand("%:p")
+		local prefix = ""
+		if ns_name ~= "" then
+			prefix = ns_name .. "_"
+		end
 
-		for _, ws in pairs(vim.lsp.buf.list_workspace_folders()) do
-			if file_path:find(ws, 1, true) == 1 then
-				ws_parts = { ws, file_path:sub(#ws + 2, -1) }
-				break
+		for name, val in pairs(eager_vars) do
+			name = prefix .. name
+			local val_type = type(val)
+			if val and val_type ~= "string" and val_type ~= "table" then
+				val = tostring(val)
 			end
+			rawset(o, name, val)
 		end
-		-- If it can't be extracted from lsp, then we use the file path
-		if not ok and not ws_parts then
-			ws_parts = { vim.fn.expand("%:p:h"), vim.fn.expand("%:p:t") }
-		end
-		vim.api.nvim_buf_set_var(0, LSP_WORSKPACE_PARTS, ws_parts)
 	end
-	return ws_parts
+	return o
 end
 
-function lazy_vars.RELATIVE_FILEPATH() -- The relative (to the opened workspace or folder) file path of the current document
-	return buf_to_ws_part()[2]
-end
+local builtin_ns_names = vim.inspect(vim.tbl_keys(builtin_namespace.builtin_ns))
+
+local function _env_namespace(name, opts)
+	assert(
+		opts and type(opts) == "table",
+		("Your namespace '%s' has to be a table"):format(name)
```

Comment:
`namespace` -> `opts`? (should be understandable since the second arg is called just that in the documentation)